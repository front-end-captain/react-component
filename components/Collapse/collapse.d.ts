import * as React from "react";

import CollapsePanel from "./collapsePanel";

export interface CollapseProps {
  style?: React.CSSProperties;
  className?: string;
  prefixCls?: string;

  activeKey?: Array<string> | string;
  defaultActiveKey?: Array<string>;

  // 手风琴模式
  accordion?: boolean;

  // 销毁隐藏的折叠面板
  // destroyInactivePanel?: boolean;
  onChange?: (key: string | string[]) => void;
  bordered?: boolean;

  expandIcon?: React.ReactNode | ((panelProps: CollapseProps) => React.ReactNode);
}

export interface CollapseState {
  activeKey: Array<string>;
}

export default class Collapse extends React.Component<CollapseProps, any> {
  static Panel: typeof CollapsePanel;
}
