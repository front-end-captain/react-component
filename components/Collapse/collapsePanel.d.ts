import * as React from 'react';

export interface CollapsePanelProps {
    id: string;
    className?: string;
    style?: React.CSSProperties;
    prefixCls: string;
    key: string;

    // 面板头内容
    header: React.ReactNode;
    disabled?: boolean;
    showArrow?: boolean;

    // 面板隐藏时是否渲染 DOM 结构
    // forceRender?: boolean;
    isActive: boolean;
    // destroyInactivePanel: boolean;
    arrordion: boolean;
    expandIcon?: React.ReactNode | ((panelProps: CollapseProps) => React.ReactNode);
}
export default class CollapsePanel extends React.Component<CollapsePanelProps, {}> {}
