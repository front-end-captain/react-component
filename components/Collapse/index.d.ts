import Collapse from "./collapse";

export { CollapseProps } from "./collapse";
export { CollapsePanelProps } from "./collapsePanel";

export default Collapse;
