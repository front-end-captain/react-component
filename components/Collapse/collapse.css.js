import styled from "styled-components";

const CollapseWrapper = styled.div`
  font-size: 14px;
  line-height: 1.5;
  color: rgba(0, 0, 0, 0.65);
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  list-style: none;
  background-color: #fafafa;
  border-radius: 4px;
  border: 1px solid #d9d9d9;
  border-bottom: 0;

  &.${(props) => props.prefixCls}-wrapper-no-bordered {
    border: none;
    background-color: #fff;

    .${(props) => props.prefixCls}-panel-content {
      border-top: none;
      background-color: transparent;
    }
  }
`;

export default CollapseWrapper;
