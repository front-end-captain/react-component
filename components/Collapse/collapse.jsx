import React, { Component, Children, Fragment, cloneElement } from "react";
import PropTypes from "prop-types";
import { isFragment } from "react-is";
import classNames from "classnames";
import { TransitionGroup } from "react-transition-group";

import CollapseWrapper from "./collapse.css.js";

const toArray = (activeKey) => {
  let currentActiveKey = activeKey;
  if (!Array.isArray(currentActiveKey)) {
    currentActiveKey = currentActiveKey ? [currentActiveKey] : [];
  }
  return currentActiveKey;
};

class Collapse extends Component {
  static defaultProps = {
    prefixCls: "sky-collapse",
    accordion: false,
    bordered: true,
    onChange: () => {},
    defaultActiveKey: ["1"],
  };

  static propTypes = {
    style: PropTypes.objectOf(PropTypes.string),
    className: PropTypes.string,
    prefixCls: PropTypes.string,
    activeKey: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)]),
    defaultActiveKey: PropTypes.arrayOf(PropTypes.string),
    accordion: PropTypes.bool,
    destroyInactivePanel: PropTypes.bool,
    onChange: PropTypes.func,
    bordered: PropTypes.bool,
    expandIcon: PropTypes.oneOfType([PropTypes.func, PropTypes.element]),
  };

  constructor(props) {
    super(props);

    const { activeKey, defaultActiveKey } = props;
    let currentActiveKey = defaultActiveKey;

    if ("activeKey" in props) {
      currentActiveKey = activeKey;
    }

    this.state = {
      activeKey: toArray(currentActiveKey),
    };
  }

  handleChild = (child, index) => {
    if (!child) {
      return null;
    }

    const { activeKey } = this.state;
    const key = child.key || String(index);
    const { prefixCls, destroyInactivePanel, accordion, expandIcon } = this.props;
    const { header, className: headerClassName, disabled, style } = child.props;

    let isActive;
    if (accordion) {
      isActive = activeKey[0] === key;
    } else {
      isActive = activeKey.includes(key);
    }

    const props = {
      key,
      header,
      isActive,
      prefixCls,
      style,
      destroyInactivePanel,
      accordion,
      expandIcon,
      panelKey: key,
      className: headerClassName,
      children: child.props.children,
      onItemHeaderClick: disabled ? null : this.onItemHeaderClick,
    };

    return cloneElement(child, props);
  };

  getCollapseItem = () => {
    const { children } = this.props;
    const realChildren = isFragment(children) ? children.props.children : children;
    const newChildren = Children.map(realChildren, this.handleChild);

    if (isFragment(children)) {
      return <Fragment>{newChildren}</Fragment>;
    }

    return newChildren;
  };

  onItemHeaderClick = (key) => {
    const { activeKey } = this.state;
    const { accordion } = this.props;

    let newActiveKey;
    if (accordion) {
      newActiveKey = activeKey[0] === key ? [] : [key];
    } else {
      const index = activeKey.indexOf(key);
      newActiveKey = Array.from(activeKey);
      if (index > -1) {
        newActiveKey.splice(index, 1);
      } else {
        newActiveKey.push(key);
      }
    }

    this.setActiveKey(newActiveKey);
  };

  setActiveKey = (newActiveKey) => {
    if (!("activeKey" in this.props)) {
      this.setState({ activeKey: newActiveKey });
    }

    const { onChange, accordion } = this.props;
    onChange(accordion ? newActiveKey[0] : newActiveKey);
  };

  render() {
    const { prefixCls, className, style, bordered } = this.props;
    const collapseWrapperClassName = classNames(`${prefixCls}-wrapper`, className, {
      [`${prefixCls}-wrapper-no-bordered`]: !bordered,
    });

    return (
      <CollapseWrapper prefixCls={prefixCls} className={collapseWrapperClassName} style={style}>
        <TransitionGroup component={null}>{this.getCollapseItem()}</TransitionGroup>
      </CollapseWrapper>
    );
  }
}

export default Collapse;
