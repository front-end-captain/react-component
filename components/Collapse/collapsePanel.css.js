import styled from "styled-components";

const PanelWrapper = styled.div`
  border-bottom: 1px solid #d9d9d9;

  .${(props) => props.prefixCls}-panel-header {
    line-height: 22px;
    padding: 12px 0 12px 40px;
    color: rgba(0, 0, 0, 0.85);
    cursor: pointer;
    position: relative;
    transition: all 0.3s;

    .${(props) => props.prefixCls}-panel-header-icon {
      font-style: normal;
      vertical-align: -0.125em;
      text-align: center;
      text-transform: none;
      text-rendering: optimizeLegibility;
      -webkit-font-smoothing: antialiased;
      -moz-osx-font-smoothing: grayscale;
      font-size: 14px;
      position: absolute;
      display: inline-block;
      line-height: 46px;
      top: 50%;
      margin-top: 2px;
      transform: translateY(-50%);
      left: 16px;
    }
  }

  .${(props) => props.prefixCls}-panel-content {
    overflow: hidden;
    color: rgba(0, 0, 0, 0.65);
    background-color: #fff;
    border-top: 1px solid #d9d9d9;
    text-indent: 2em;
    padding: 16px;
    transition: all 0.2s;
  }

  .${(props) => props.prefixCls}-panel-content-inactive {
    height: 0;
    border: none;
    padding-top: 0;
    padding-bottom: 0;
  }

  &.${(props) => props.prefixCls}-panel-disabled {
    cursor: not-allowed;
    color: rgba(0, 0, 0, 0.25);

    .${(props) => props.prefixCls}-panel-header-icon {
      cursor: not-allowed;
      color: rgba(0, 0, 0, 0.25);
    }
  }
`;

export default PanelWrapper;
