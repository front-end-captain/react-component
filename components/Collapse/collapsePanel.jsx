import React, { Component, createElement } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";

import { DownIcon } from "./../Icon/index.jsx";

import PanelWrapper from "./collapsePanel.css.js";

class CollapsePanel extends Component {
  static defaultProps = {
    showArrow: true,
    isActive: false,
    // destroyInactivePanel: false,
    onItemHeaderClick: () => {},
    // forceRender: false,
  };

  static propTypes = {
    className: PropTypes.string,
    id: PropTypes.string,
    prefixCls: PropTypes.string,
    header: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.element]),
    showArrow: PropTypes.bool,
    isActive: PropTypes.bool,
    onItemHeaderClick: PropTypes.func,
    style: PropTypes.objectOf(PropTypes.string),
    // destroyInactivePanel: PropTypes.bool,
    disabled: PropTypes.bool,
    accordion: PropTypes.bool,
    // forceRender: PropTypes.bool,
    expandIcon: PropTypes.func,
    panelKey: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  };

  shouldComponentUpdate() {
    return true;
  }

  handleHeaderClick = () => {
    const { onItemHeaderClick, panelKey } = this.props;
    if (typeof onItemHeaderClick === "function") {
      onItemHeaderClick(panelKey);
    }
  };

  render() {
    const {
      className,
      id,
      style,
      prefixCls,
      header,
      children,
      isActive,
      showArrow,
      // destroyInactivePanel,
      disabled,
      accordion,
      // forceRender,
      expandIcon,
    } = this.props;

    const panelClassName = classNames(`${prefixCls}-panel`, className, {
      [`${prefixCls}-panel-active`]: isActive,
      [`${prefixCls}-panel-disabled`]: disabled,
    });
    const panelContentBoxClassName = classNames(`${prefixCls}-panel-content`, {
      [`${prefixCls}-panel-content-active`]: isActive,
      [`${prefixCls}-panel-content-inactive`]: !isActive,
    });

    let arrowIcon = null;
    if (showArrow && typeof expandIcon === "function") {
      arrowIcon = createElement(expandIcon, { ...this.props });
    }

    return (
      <PanelWrapper className={panelClassName} prefixCls={prefixCls} style={style} id={id}>
        <div className={`${prefixCls}-panel-header`} onClick={this.handleHeaderClick}>
          <span className={`${prefixCls}-panel-header-icon`}>
            {showArrow &&
              (arrowIcon || (
                <DownIcon
                  style={{ transform: `rotate(${isActive ? -90 : 0}deg)`, transition: "all 0.2s" }}
                />
              ))}
          </span>
          {header}
        </div>
        <div className={panelContentBoxClassName} role={accordion ? "tabpanel" : null}>
          {children}
        </div>
      </PanelWrapper>
    );
  }
}

export default CollapsePanel;
