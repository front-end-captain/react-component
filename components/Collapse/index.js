import Collapse from "./collapse.jsx";
import CollapsePanel from "./collapsePanel.jsx";

Collapse.Panel = CollapsePanel;

export default Collapse;
