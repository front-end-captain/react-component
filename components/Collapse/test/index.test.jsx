import React from "react";
import { render } from "enzyme";
import toJson from "enzyme-to-json";

import Collapse from "./../index.js";

const { Panel } = Collapse;

describe("Panel", () => {
  it("should render some Collapse components", () => {
    const wrpaper = render(
      <div>
        <Collapse>
          <Panel header="title" key="1">
            some text
          </Panel>
          <Panel header="title" key="2">
            some text
          </Panel>
          <Panel header="title" key="3">
            some text
          </Panel>
        </Collapse>
        <Collapse accordion>
          <Panel header="title" key="1">
            some text
          </Panel>
          <Panel header="title" key="2">
            some text
          </Panel>
          <Panel header="title" key="3">
            some text
          </Panel>
        </Collapse>
        <Collapse>
          <Panel header="This is panel header 1" key="1">
            <Collapse defaultActiveKey={["1"]}>
              <Panel header="This is panel nest panel" key="1">
                some text
              </Panel>
            </Collapse>
          </Panel>
          <Panel header="This is panel header 2" key="2">
            some text
          </Panel>
          <Panel header="This is panel header 3" key="3">
            some text
          </Panel>
        </Collapse>
      </div>,
    );

    expect(toJson(wrpaper)).toMatchSnapshot();
  });
});
