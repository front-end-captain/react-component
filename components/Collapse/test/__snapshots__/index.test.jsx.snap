// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Panel should render some Collapse components 1`] = `
<div>
  <div
    class="sky-collapse-wrapper collapsecss__CollapseWrapper-dsnzp-0 fkgxCj"
  >
    <div
      class="sky-collapse-panel sky-collapse-panel-active collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(-90deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        title
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-active"
      >
        some text
      </div>
    </div>
    <div
      class="sky-collapse-panel collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(0deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        title
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-inactive"
      >
        some text
      </div>
    </div>
    <div
      class="sky-collapse-panel collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(0deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        title
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-inactive"
      >
        some text
      </div>
    </div>
  </div>
  <div
    class="sky-collapse-wrapper collapsecss__CollapseWrapper-dsnzp-0 fkgxCj"
  >
    <div
      class="sky-collapse-panel sky-collapse-panel-active collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(-90deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        title
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-active"
        role="tabpanel"
      >
        some text
      </div>
    </div>
    <div
      class="sky-collapse-panel collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(0deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        title
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-inactive"
        role="tabpanel"
      >
        some text
      </div>
    </div>
    <div
      class="sky-collapse-panel collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(0deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        title
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-inactive"
        role="tabpanel"
      >
        some text
      </div>
    </div>
  </div>
  <div
    class="sky-collapse-wrapper collapsecss__CollapseWrapper-dsnzp-0 fkgxCj"
  >
    <div
      class="sky-collapse-panel sky-collapse-panel-active collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(-90deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        This is panel header 1
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-active"
      >
        <div
          class="sky-collapse-wrapper collapsecss__CollapseWrapper-dsnzp-0 fkgxCj"
        >
          <div
            class="sky-collapse-panel sky-collapse-panel-active collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
          >
            <div
              class="sky-collapse-panel-header"
            >
              <span
                class="sky-collapse-panel-header-icon"
              >
                <svg
                  fill="none"
                  height="1em"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  style="transform:rotate(-90deg);transition:all 0.2s"
                  viewBox="0 0 24 24"
                  width="1em"
                >
                  <polyline
                    points="6 9 12 15 18 9"
                  />
                </svg>
              </span>
              This is panel nest panel
            </div>
            <div
              class="sky-collapse-panel-content sky-collapse-panel-content-active"
            >
              some text
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sky-collapse-panel collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(0deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        This is panel header 2
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-inactive"
      >
        some text
      </div>
    </div>
    <div
      class="sky-collapse-panel collapsePanelcss__PanelWrapper-ay2z5j-0 gNuQND"
    >
      <div
        class="sky-collapse-panel-header"
      >
        <span
          class="sky-collapse-panel-header-icon"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="transform:rotate(0deg);transition:all 0.2s"
            viewBox="0 0 24 24"
            width="1em"
          >
            <polyline
              points="6 9 12 15 18 9"
            />
          </svg>
        </span>
        This is panel header 3
      </div>
      <div
        class="sky-collapse-panel-content sky-collapse-panel-content-inactive"
      >
        some text
      </div>
    </div>
  </div>
</div>
`;
